src/errors.rs

```diff
@@ -47,6 +50,145 @@ impl Display for ErrorType {
+    fn from(value: Result<T, ConvertedError>) -> Self {
+        match value {
+            Ok(v) => JsResult::Ok(v),
+            Err(e) => JsResult::Error(e),
+        }
+    }
+}
+
+impl<T> ToNapiValue for JsResult<T>
+where
+    T: ToNapiValue,
+{
+    unsafe fn to_napi_value(
+        env: napi::sys::napi_env,
+        val: Self,
+    ) -> napi::Result<napi::sys::napi_value> {
+        match val {
+            JsResult::Ok(val) => unsafe { T::to_napi_value(env, val) },
+            JsResult::Error(error) => {
+                // # Safety - Valid pointer to napi env is provided by the caller
+                let v = unsafe { ConvertedError::to_napi_value(env, error) }?;
+                // # Safety - We are creating Unknown from valid napi env (validity ensured by the caller) and value (created just above)
+                let err = unsafe { Unknown::from_raw_unchecked(env, v) };
+
+                Err(err.into())
+            }
+            JsResult::NapiError(err) => Err(err),
+        }
+    }
+}
+
+pub struct ConvertedError {
+    msg: String,
+    name: String,
+}
+
+impl<T> From<T> for ConvertedError
+where
+    T: Error,
+{
+    fn from(value: T) -> Self {
+        ConvertedError {
+            msg: value.to_string(),
+            // We obtain the error class name.
+            // Because this is never empty, we will never fail to get the last element
+            // The split is done to cope with the fact that the type name
+            // may return type with full package or without.
+            //
+            // Why using such obscure method?
+            // Rust driver maintainer did not agree to add a code that would allow
+            // systematic conversion from the errors as they are into their names and so on...
+            //
+            // For this reason, we also only expose the error class name,
+            // and not the name of the specific kind, or the inner values of those errors.
+            name: std::any::type_name::<T>()
+                .split(":")
+                .last()
+                .expect("Expected to obtain error name")
+                .to_owned(),
+        }
+    }
+}
+
+impl ToNapiValue for ConvertedError {
+    unsafe fn to_napi_value(
+        env: napi::sys::napi_env,
+        val: Self,
+    ) -> napi::Result<napi::sys::napi_value> {
+        let env = Env::from_raw(env);
+        let mut e = env.create_error(napi::Error::new(Status::GenericFailure, val.msg))?;
+
+        e.set_named_property("name", val.name)?;
+
+        Ok(e.raw())
+    }
+}
+
+/// Allows to run a block of code that returns Result<T, ConvertedError>,
+/// with automatic conversion to JsResult<T>. This allows to use the `?` operator,
+/// while still returning JsResult<T> from the function.
+macro_rules! with_custom_error {
+    ($code: block) => {
+        async {
+            let c: Result<_, ConvertedError> = async { $code }.await.into();
+
+            c.into()
+        }
+        .await
+    };
+}
```

Comment:
:100: Good idea! But does this really need to be a macro? Can't we make this a function?